pkg/storegateway/batch_series_parallel.go

```diff
@@ -0,0 +1,137 @@
+package storegateway
+
+import (
+	"io"
+
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/storage"
+	"github.com/prometheus/prometheus/tsdb/chunks"
+	"github.com/thanos-io/thanos/pkg/pool"
+
+	"github.com/grafana/mimir/pkg/storegateway/indexheader"
+	"github.com/grafana/mimir/pkg/storegateway/storepb"
+)
+
+// batchSeriesSet will implement storepb.SeriesSet
+//
+// Implementation will be similar to existing batchedSeriesSet
+// A batchSeriesSet is created and used in BucketStore.Series
+type batchSeriesSet interface {
+	storage.SeriesSet
+
+	seriesWatcher
+}
+
+// batchSeriesSetFactory is used in BucketStore.Series
+type batchSeriesSetFactory func(
+	// postings can also become a postings provider, but for now we fetch all postings in one go.
+	// It can also be calculated by the seriesSet itself. For simplicity of this interface it's just provided.
+	postings []storage.SeriesRef,
+	batcher batcher,
+	batchFactory batchFactory, // the batchedSeriesSet creates its own batches and passes them to be populated
+) batchSeriesSet
+
+type seriesWatcher interface {
+	// seriesDisposed is a callback when a series can be safely discarded.
+	// This method will be called after BucketStore.Series sends the series over gRPC.
+	//
+	// Once the batchSeriesSet receives a seriesDisposed call, it can look up
+	// whether the disposed series is beyond an existing batch.
+	// If that's the case it can close that batch.
+	//
+	// This relies on BucketStore.Series iterating through the series in a sorted manner,
+	// because batchSeriesSet will infer that if series X is disposed, then so must be series X-1 and X-2
+	//
+	// The thing that invokes seriesDisposed will be a tracker somewhere in BucketStore.Series
+	// or a top-level storepb.SeriesSet implementation, which calls
+	seriesDisposed(seriesID storage.SeriesRef)
+}
+
+// batcher fetches batches in parallel while respecting the global quotas by its quotaManager
+// batcher is created in BucketStore.Series
+// batcher can also potentially offload batches to disk as long as batch keeps being an interface
+type batcher interface {
+	populate(b batch)
+}
+
+// batcherFactory is used in BucketStore.Series
+type batcherFactory func(
+	quotas quotaManager,
+	filter shardingFilter, // since the batcher is the first to know the labelset, it's the first to be able to filter out un-owned series
+	seriesReader seriesReader,
+	chunkReader chunkReader,
+	minTime, maxTime int,
+	skipChunks bool,
+	// maybe also ULID and tenant ID
+) batcher
+
+type batch interface {
+	io.Closer // takes care of freeing resources
+
+	postings() []storage.SeriesRef
+	series() []seriesEntry                                           // to be used by the batchedSeriesSet to emulate a storepb.SeriesSet
```

Comment:
Looks weird to me having both `postings()` and `series()`. Postings are lower reference to a series, so why not just having `series()` where the returned value include the `storage.SeriesRef` too?

Reply:
yeah we can have that. My concern was that we'd add more fields to the existing `seriesEntry` and increase its memory footprint; this may not be material though.

The root reason for having them separated is that we first have only the postings without any series or chunks. And the batcher needs the postings to start populating the batch.

I was considering a batchBuilder which will have only `postings()` and the batch will only have `series()`. The batchFactory will return a batchBuilder and the batcher will take a batchBuilder and return a batch